[DOC] Topic operator usage recommendations (#1738)
strimzi/strimzi-kafka-operator: documentation/book/ref-topic-operator-usage-recommendation.adoc

[id='topic-operator-usage-recommendations-{context}']
= Topic Operator usage recommendations

When working with topics, be consistent and always operate on either `KafkaTopic` resources or topics directly. Avoid routinely switching between both methods for a given topic.

Use topic names that reflect the nature of the topic, and remember that names cannot be changed later.

If creating a name directly for a topic, use a name that is a valid {KubernetesName} resource name, otherwise the Topic Operator will need to modify the name when creating the corresponding `KafkaTopic`.

.Kafka topic naming conventions

Kafka and {KubernetesName} impose their own validation rules for the naming of `KafkaTopic` and `KafkaTopic.metadata.name` respectively.

> So there are valid naming conventions for each which are invalid in the other.

Reviewer: Can we link to the rules directly here and not only at the end in additional resources section?

Reviewer 2: I hate `Additional Resources`. Who thinks to look there when reading the documentation as a web page? We expect such references to be hyperlinks inline with the text.

Reviewer: Right. I personally don't mind having them there. But the links should be always also in the text.

Author: I guess the idea with Additional resources is to link to content outside the current assembly. And to encourage the reader to concentrate on the content in front of them without navigating away. I agree that it can seem a bit cluttered in the web format. I don't think we need links in the content and an Additional resources in this context, so removed Additional resources.

Reviewer: Well, I think that if you don't want the link there then you need to briefly explain the rules in the text. Because it is not useful to treat the reader as _there are some restrictions but we will not tell you which_. The additional resources are not useful here because the user will not have them when he reads this sentence.

The `spec.topicName` property provides a way to create a name for a topic in Kafka that would be a valid name for `KafkaTopic` in {KubernetesName}.

The `spec.topicName` property inherits Kafka naming validation rules:

* The name must not be longer than 249 characters
* Valid characters for Kafka topics are ASCII alphanumerics, ‘.’, ‘_’, and ‘-‘

> `spec.topicName` must not be changed. For example, `KafkaTopic.metadata.name=A with KafkaTopic.spec.name=name-1` cannot be changed to `KafkaTopic.metadata.name=A with KafkaTopic.spec.topicName=name-2`.

Reviewer: ```suggestion
`spec.topicName` must not be changed. For example, `KafkaTopic.metadata.name=A with KafkaTopic.spec.name=topicName-1` cannot be changed to `KafkaTopic.metadata.name=A with KafkaTopic.spec.topicName=name-2`.
```

[id='topic-operator-usage-recommendations-{context}']
= Topic Operator usage recommendations

When working with topics, be consistent and always operate on either `KafkaTopic` resources or topics directly. Avoid routinely switching between both methods for a given topic.

Use topic names that reflect the nature of the topic, and remember that names cannot be changed later.

> If creating a name directly for a topic, use a name that is a valid {KubernetesName} resource name, otherwise the Topic Operator will need to modify the name when creating the corresponding `KafkaTopic`.

Reviewer 2: Let's be clear that:

* we create topics, which have a name, rather than "creating topic names".
* the name is only modified on the Kube side.

```suggestion
If creating a topic in Kafka, use a name that is a valid {KubernetesName} resource name, otherwise the Topic Operator will need to create the corresponding `KafkaTopic` with a different name which obeys the {KubernetesName} rules.
```

.Kafka topic naming conventions

> Kafka and {KubernetesName} impose their own validation rules for the naming of `KafkaTopic` and `KafkaTopic.metadata.name` respectively.

Reviewer 2: Kafka has no notion of `KafkaTopic`, so this is confusing worded, how about:

```suggestion
Kafka and {KubernetesName} impose their own validation rules for the naming of topics in Kafka and `KafkaTopic.metadata.name` respectively.
```

> So there are valid naming conventions for each which are invalid in the other.

Reviewer 2: This isn't really about naming conventions, as such, more about names themselves.
And the "So" doesn't work because the preceding text hasn't actually explained that the rules are. 

```suggestion
There are valid names for each which are invalid in the other.
```

> The `spec.topicName` property provides a way to create a name for a topic in Kafka that would be a valid name for `KafkaTopic` in {KubernetesName}, link:https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/identifiers.md[(according to the recommendations for identifiers and names in {KubernetesName})^].

Reviewer 2: ```suggestion
The `spec.topicName` property provides a way to create a topic in Kafka with a name that would be invalid for the `KafkaTopic` in {PlatformProductName}, link:https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/identifiers.md[(according to the recommendations for identifiers and names in {KubernetesName})^].
```

The `spec.topicName` property inherits Kafka naming validation rules:

* The name must not be longer than 249 characters

> * Valid characters for Kafka topics are ASCII alphanumerics, ‘.’, ‘_’, and ‘-‘

Reviewer 2: Strictly speaking it can't be `.` or `..` either. I suppose we should mention that.

Author: * The name cannot be `.` or `..`, though `.` can be used in a name, such as `exampleTopic.` or `.exampleTopic`

`spec.topicName` must not be changed.

For example:

[source,yaml]
----

> KafkaTopic.metadata.name=A with KafkaTopic.spec.name=topicName-1

Reviewer 2: Let's do the example resources which will be familiar to the reader

```suggestion
kind: KafkaTopic
metadata:
  name: topic-name-1
spec:
  topicName: topicName-1 # Upper case is invalid in {KubernetesName}
  # ...
```

----

cannot be changed to

[source,yaml]
----

> KafkaTopic.metadata.name=A with KafkaTopic.spec.topicName=name-2

Reviewer 2: ```suggestion
kind: KafkaTopic
metadata:
  name: topic-name-1
spec:
  topicName: name-2
  # ...
```

[id='topic-operator-usage-recommendations-{context}']
= Topic Operator usage recommendations

When working with topics, be consistent and always operate on either `KafkaTopic` resources or topics directly. Avoid routinely switching between both methods for a given topic.

Use topic names that reflect the nature of the topic, and remember that names cannot be changed later.

> If creating a topic in Kafka, use a name that is a valid {KubernetesName} resource name, otherwise the Topic Operator will need to create the corresponding `KafkaTopic` with a different name which obeys the {KubernetesName} rules.

Reviewer 3: ```suggestion
If creating a topic in Kafka, use a name that is a valid {KubernetesName} resource name, otherwise the Topic Operator will need to create the corresponding `KafkaTopic` with a name that conforms to the {KubernetesName} rules.
```

.Kafka topic naming conventions

Kafka and {KubernetesName} impose their own validation rules for the naming of topics in Kafka and `KafkaTopic.metadata.name` respectively.
There are valid names for each which are invalid in the other.

> 

Reviewer 3: Recommendations for identifiers and names in {KubernetesName}  are outlined in link:https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/identifiers.md[Identifiers and Names in Kubernetes^] community article.

> The `spec.topicName` property provides a way to create a topic in Kafka with a name that would be invalid for the `KafkaTopic` in {PlatformProductName}, link:https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/identifiers.md[(according to the recommendations for identifiers and names in {KubernetesName})^].

Reviewer 3: "provides a way" almost sounds like a recommendation. I suggest something along the following lines:

Using the `spec.topicName` property, it is possible to create a valid topic in Kafka with a name that is invalid for the `KafkaTopic` in {PlatformProductName}.

I feel the link to the Kubernetes article should be introduced earlier.

Author: Better. I moved the link up too.

The `spec.topicName` property inherits Kafka naming validation rules:

> 

Reviewer 3: Use full stops at the end of list items that are complete sentences.

* The name must not be longer than 249 characters

> * Valid characters for Kafka topics are ASCII alphanumerics, ‘.’, ‘_’, and ‘-‘

Reviewer 3: Enclose the symbols using backticks.

Author: I did this, but they're hard to see on my local build.

Reviewer 2: Name the characters for the avoidance of doubt:

```suggestion
* Valid characters for Kafka topics are ASCII alphanumerics, ‘.’ (full stop/period), ‘_’ (low line/underscore), and ‘-‘ (hyphen-minus)
```

(I know full stop, low line and hyphen-minus might seem like unconventional names, but they're actually the names defined in Unicode).

> * The name cannot be `.` or `..`, though `.` can be used in a name, such as `exampleTopic.` or `.exampleTopic`

Reviewer 3: ```suggestion
* The name cannot be `.` or `..`, although `.` can be used as part of a name, such as `exampleTopic.` or `.exampleTopic`
```

`spec.topicName` must not be changed.

For example:

[source,yaml]
----
kind: KafkaTopic
metadata:
  name: topic-name-1
spec:
  topicName: topicName-1 # Upper case is invalid in {KubernetesName}
  # ...
----

cannot be changed to

[source,yaml]
----
kind: KafkaTopic
metadata:
  name: topic-name-1
spec:
  topicName: name-2
  # ...
----

> NOTE: Some Kafka client applications, such as Kafka Streams, can create topics in Kafka programmatically. When those topics have a name that is an invalid {KubernetesName} resource name, the Topic Operator gives them a valid name based on the Kafka name. Invalid characters are replaced and a hash appended to the name.

Reviewer 3: ```suggestion
NOTE: Some Kafka client applications, such as Kafka Streams, can create topics in Kafka programmatically. If those topics have names that are invalid {KubernetesName} resource names, the Topic Operator gives them valid names based on the Kafka names. Invalid characters are replaced and a hash is appended to the name.
```
